src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java

```diff
@@ -1,437 +1,58 @@
+  private final Collection<OddsCalculator> workers = new HashSet<>();
 
-  private int currentThreads = MAX_THREADS;
-  private final ExecutorService executor;
-  private final List<OddsCalculator> workers = new CopyOnWriteArrayList<>();
-  // do not let calc be set up til data is set
-  private volatile boolean isDataSet = false;
-  // do not let calc start until it is set
-  private volatile boolean isCalcSet = false;
-  // shortcut everything if we are shutting down
-  private volatile boolean isShutDown = false;
-  // shortcut setting of previous game data if we are trying to set it to a new one, or shutdown
-  private volatile int cancelCurrentOperation = 0;
-  // do not let calcing happen while we are setting game data
-  private final CountUpAndDownLatch latchSetData = new CountUpAndDownLatch();
-  // do not let setting of game data happen multiple times while we offload creating workers and copying data to a
-  // different thread
-  private final CountUpAndDownLatch latchWorkerThreadsCreation = new CountUpAndDownLatch();
-
-  // do not let setting of game data happen at same time
-  private final Object mutexSetGameData = new Object();
-  // do not let multiple calculations or setting calc data happen at same time
-  private final Object mutexCalcIsRunning = new Object();
-  private final List<OddsCalculatorListener> listeners = new ArrayList<>();
-
-  public ConcurrentOddsCalculator(final String threadNamePrefix) {
-    executor = Executors.newFixedThreadPool(MAX_THREADS,
-        new DaemonThreadFactory(true, threadNamePrefix + " ConcurrentOddsCalculator Worker"));
-    logger.fine("Initialized executor thread pool with size: " + MAX_THREADS);
-  }
-
-  @Override
-  public void setGameData(final GameData data) {
-    // increment so that a new calc doesn't take place (since they all wait on this latch)
-    latchSetData.increment();
-    // cancel any current setting of data
-    --cancelCurrentOperation;
-    // cancel any existing calcing (it won't stop immediately, just quicker)
-    cancel();
-    synchronized (mutexSetGameData) {
-      try {
-        // since setting data takes place on a different thread, this is our token. wait on it since
-        latchWorkerThreadsCreation.await();
-        // we could have exited the synchronized block already.
-      } catch (final InterruptedException e) {
-        Thread.currentThread().interrupt();
-      }
-      cancel();
-      isDataSet = false;
-      isCalcSet = false;
-      if (data == null || isShutDown) {
-        workers.clear();
-        ++cancelCurrentOperation;
-        // allow calcing and other stuff to go ahead
-        latchSetData.countDown();
-      } else {
-        ++cancelCurrentOperation;
-        // increment our token, so that we can set the data in a different thread and return from this one
-        latchWorkerThreadsCreation.increment();
-        executor.submit(() -> createWorkers(data));
-      }
-    }
-  }
+  private boolean cancelled = false;
```

Comment:
This field should probably be `volatile` because it is read/written from/to by different threads.

Reply:
cde48b2